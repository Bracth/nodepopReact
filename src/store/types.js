export const AUTH_LOGIN_SUCCESS = "AUTH_LOGIN_SUCESS";
export const AUTH_LOGOUT_SUCCESS = "AUTH_LOGOUT_SUCCESS";
export const ADVERTS_LOADED = "ADVERTS_LOADED";
