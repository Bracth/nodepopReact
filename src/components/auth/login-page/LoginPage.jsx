import {useState} from "react"
import { login } from "../service";

function LoginPage() {
    
    const [credentials, setCredentials] = useState({
        email: "",
        password: "",
        remember: false
    });
    
    const {email, password, remember} = credentials
    
    const handleChange = (event) => {
        setCredentials(credentials => ({
            ...credentials,
            [event.target.name]: 
                event.target.type === "checkbox" 
                    ? event.target.checked
                    : event.target.value,
        }))
    }
    
    const handleSubmit = async (event) => {
        event.preventDefault();
        try {
        await login(credentials);
        } catch (error) {
            console.log(error.message)
        }
    }
    
    return <main>
        <h1>Login to Nodepop</h1>
        <form onSubmit={handleSubmit}>
            <input type="text" placeholder="email" name="email" value={email} onChange={handleChange}/>
            <input type="password" placeholder="password" name="password" value={password} onChange={handleChange}/>
            <input type="checkbox" name="remember" value="remember" checked={remember} onChange={handleChange}/>
            <label htmlFor="remember">remember</label>
            <button>Login</button>
        </form>
    </main>
}

export default LoginPage;