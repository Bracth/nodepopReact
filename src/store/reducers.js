import {
  ADVERTS_LOADED_REQUEST,
  ADVERTS_LOADED_SUCCESS,
  ADVERTS_LOADED_FAILURE,
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT_SUCCESS,
  UI_RESET_ERROR,
  ADVERT_LOADED_SUCCESS,
  ADVERT_LOADED_REQUEST,
  ADVERT_CREATED_REQUEST,
  ADVERT_CREATED_SUCCESS,
  ADVERT_LOADED_FAILURE,
  ADVERT_CREATED_FAILURE,
  ADVERT_TAGS_LOADED_REQUEST,
  ADVERT_TAGS_LOADED_SUCCESS,
  ADVERT_TAGS_LOADED_FAILURE,
  ADVERT_DELETED_SUCCESS,
  ADVERT_DELETED_REQUEST,
  ADVERT_DELETED_FAILURE,
  AUTH_REGISTER_REQUEST,
  AUTH_REGISTER_SUCCESS,
  AUTH_REGISTER_FAILURE,
} from "./types";

export const defaulState = {
  auth: true,
  adverts: {
    isLoaded: false,
    data: [],
  },
  tags: {
    isLoaded: false,
    data: [],
  },
  ui: {
    isLoading: false,
    error: null,
  },
};

export const auth = (state = defaulState.auth, action) => {
  switch (action.type) {
    case AUTH_LOGIN_SUCCESS:
      return true;
    case AUTH_LOGOUT_SUCCESS:
      return false;
    default:
      return state;
  }
};

export const adverts = (state = defaulState.adverts, action) => {
  switch (action.type) {
    case ADVERTS_LOADED_SUCCESS:
      return { ...state, isLoaded: true, data: action.payload };
    case ADVERT_LOADED_SUCCESS:
      return { ...state, data: [...state.data, action.payload] };
    case ADVERT_CREATED_SUCCESS:
      return { ...state, data: [...state.data, action.payload] };
    case ADVERT_DELETED_SUCCESS:
      return {
        ...state,
        data: state.data.filter((advert) => advert.id !== action.payload),
      };
    default:
      return state;
  }
};

export const tags = (state = defaulState.tags, action) => {
  switch (action.type) {
    case ADVERT_TAGS_LOADED_SUCCESS:
      return { isLoaded: true, data: action.payload };
    default:
      return state;
  }
};

export const ui = (state = defaulState.ui, action) => {
  switch (action.type) {
    case AUTH_LOGIN_REQUEST:
    case AUTH_REGISTER_REQUEST:
    case ADVERTS_LOADED_REQUEST:
    case ADVERT_LOADED_REQUEST:
    case ADVERT_CREATED_REQUEST:
    case ADVERT_TAGS_LOADED_REQUEST:
    case ADVERT_DELETED_REQUEST:
      return { ...state, isLoading: true, error: null };
    case AUTH_LOGIN_SUCCESS:
    case AUTH_REGISTER_SUCCESS:
    case ADVERTS_LOADED_SUCCESS:
    case ADVERT_LOADED_SUCCESS:
    case ADVERT_CREATED_SUCCESS:
    case ADVERT_TAGS_LOADED_SUCCESS:
    case ADVERT_DELETED_SUCCESS:
      return { ...state, isLoading: false };
    case AUTH_LOGIN_FAILURE:
    case AUTH_REGISTER_FAILURE:
    case ADVERTS_LOADED_FAILURE:
    case ADVERT_LOADED_FAILURE:
    case ADVERT_CREATED_FAILURE:
    case ADVERT_TAGS_LOADED_FAILURE:
    case ADVERT_DELETED_FAILURE:
      return { ...state, isLoading: false, error: action.payload };
    case UI_RESET_ERROR:
      return { ...state, error: null };
    default:
      return state;
  }
};
