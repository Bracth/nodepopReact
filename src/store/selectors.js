export const getIsLogged = (state) => state.auth;

export const getAdverts = (state) => state.adverts.data;

export const getDavertsIsLoaded = (state) => state.adverts.isLoaded;

export const getAdvert = (state, advertId) =>
  state.adverts.data.find((advert) => advert.id === advertId);

export const getUi = (state) => state.ui;
