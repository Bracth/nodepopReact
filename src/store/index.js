import { createStore, combineReducers, applyMiddleware } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import * as reducers from "./reducers";
import * as auth from "../components/auth/service";
import * as adverts from "../components/adverts/service";
import thunk from "redux-thunk";

const api = { auth, adverts };

const logger = (store) => (next) => (action) => {
  console.log("Before action", action, store.getState());
  const result = next(action);
  console.log("After action", action, store.getState());
  return result;
};

const timestamp = () => (next) => (action) => {
  const newAction = {
    ...action,
    meta: {
      ...action.meta,
      timestamp: new Date(),
    },
  };
  return next(newAction);
};

const configureStore = (preloadedState) => {
  const middlewares = [thunk.withExtraArgument({ api }), timestamp, logger];

  const store = createStore(
    combineReducers(reducers),
    preloadedState,
    composeWithDevTools(applyMiddleware(...middlewares))
  );

  return store;
};

export default configureStore;
