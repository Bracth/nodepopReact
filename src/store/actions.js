import { getAdvert, getDavertsIsLoaded, getTagsIsLoaded } from "./selectors";
import {
  ADVERTS_LOADED_FAILURE,
  ADVERTS_LOADED_REQUEST,
  ADVERTS_LOADED_SUCCESS,
  ADVERT_LOADED_REQUEST,
  ADVERT_LOADED_SUCCESS,
  ADVERT_LOADED_FAILURE,
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT_SUCCESS,
  UI_RESET_ERROR,
  ADVERT_CREATED_SUCCESS,
  ADVERT_CREATED_REQUEST,
  ADVERT_CREATED_FAILURE,
  ADVERT_TAGS_LOADED_SUCCESS,
  ADVERT_TAGS_LOADED_REQUEST,
  ADVERT_TAGS_LOADED_FAILURE,
} from "./types";

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
});

export const authLoginSuccess = () => ({
  type: AUTH_LOGIN_SUCCESS,
});

export const authLoginFailure = (error) => ({
  type: AUTH_LOGIN_FAILURE,
  payload: error,
  error: true,
});

export const authLogin = (credentials) => {
  return async function (dispatch, _getState, { api }) {
    try {
      dispatch(authLoginRequest());
      const response = await api.auth.login(credentials);
      if (response.status === 401) {
        throw new Error("Invalid email or password");
      }
      dispatch(authLoginSuccess());
    } catch (error) {
      dispatch(authLoginFailure(error));
    }
  };
};

export const authLogoutSucces = () => ({
  type: AUTH_LOGOUT_SUCCESS,
});

export const authLogout = () => {
  return async function (dispatch, _getState, { api }) {
    try {
      const response = await api.auth.logout();
      dispatch(authLoginSuccess());
      return response;
    } catch (error) {
      console.log(error);
    }
  };
};

export const advertsLoadedRequest = () => ({
  type: ADVERTS_LOADED_REQUEST,
});

export const advertsLoadedSuccess = (adverts) => ({
  type: ADVERTS_LOADED_SUCCESS,
  payload: adverts,
});

export const advertsLoadedFailure = (error) => ({
  type: ADVERTS_LOADED_FAILURE,
  payload: error,
  error: true,
});

export const advertsLoaded = () => {
  return async function (dispatch, getState, { api }) {
    const areAdvertsLoaded = getDavertsIsLoaded(getState());
    if (areAdvertsLoaded) return;
    try {
      dispatch(advertsLoadedRequest());
      const adverts = await api.adverts.getLastedsAdverts();
      dispatch(advertsLoadedSuccess(adverts));
    } catch (error) {
      dispatch(advertsLoadedFailure(error));
    }
  };
};

export const advertLoadedSuccess = (advert) => ({
  type: ADVERT_LOADED_SUCCESS,
  payload: advert,
});

export const advertLoadedRequest = () => ({
  type: ADVERT_LOADED_REQUEST,
});

export const advertLoadedFailure = (error) => ({
  type: ADVERT_LOADED_FAILURE,
  payload: error,
});

export const advertLoaded = (advertId) => {
  return async function (dispatch, getState, { api }) {
    const advertLoaded = getAdvert(advertId)(getState());
    if (advertLoaded) return;
    try {
      dispatch(advertLoadedRequest());
      const advert = await api.adverts.getLastedAdvert(advertId);
      dispatch(advertLoadedSuccess(advert));
    } catch (error) {
      dispatch(advertLoadedFailure(error));
    }
  };
};

export const advertTagsLoadedSuccess = (tags) => ({
  type: ADVERT_TAGS_LOADED_SUCCESS,
  payload: tags,
});

export const advertTagsLoadedRequest = () => ({
  type: ADVERT_TAGS_LOADED_REQUEST,
});

export const advertTagsLoadedFailure = (error) => ({
  type: ADVERT_TAGS_LOADED_FAILURE,
  payload: error,
});

export const advertTagsLoaded = () => {
  return async function (dispatch, getState, { api }) {
    const tagsIsLoaded = getTagsIsLoaded(getState());
    if (tagsIsLoaded) return;
    try {
      dispatch(advertTagsLoadedRequest());
      const tags = await api.adverts.getTags();
      dispatch(advertTagsLoadedSuccess(tags));
    } catch (error) {
      dispatch(advertTagsLoadedFailure(error));
    }
  };
};

export const advertCreatedSuccess = () => ({
  type: ADVERT_CREATED_SUCCESS,
});

export const advertCreatedRequest = () => ({
  type: ADVERT_CREATED_REQUEST,
});

export const advertCreatedFailure = (error) => ({
  type: ADVERT_CREATED_FAILURE,
});

export const advertCreated = (advert) => {
  return async function (dispatch, getState, { api }) {
    try {
      dispatch(advertCreatedRequest());
      const createdAdvert = await api.adverts.createAdvert(advert);
      dispatch(advertCreatedSuccess());
      return createdAdvert;
    } catch (error) {
      dispatch(advertCreatedFailure());
    }
  };
};

export const uiResetError = () => ({ type: UI_RESET_ERROR });
