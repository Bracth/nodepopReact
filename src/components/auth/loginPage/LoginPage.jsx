import { useMemo, useState } from "react";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";

import ErrorAlert from "../../error/ErrorAlert";
import Spinner from "react-bootstrap/Spinner";
import { useDispatch, useSelector } from "react-redux";
import { authLogin } from "../../../store/actions";
import { getUi } from "../../../store/selectors";

function LoginPage() {
  const [credentials, setCredentials] = useState({
    email: "",
    password: "",
    remember: false,
  });

  const { email, password, remember } = credentials;

  const dispatch = useDispatch();
  const { isLoading, error } = useSelector(getUi);

  const handleChange = ({ target: { value, name, type, checked } }) => {
    setCredentials((credentials) => ({
      ...credentials,
      [name]: type === "checkbox" ? checked : value,
    }));
  };

  const handleSubmit = async (event) => {
    event.preventDefault();
    dispatch(authLogin(credentials));
  };

  const disabledButton = useMemo(() => {
    return !email || !password || isLoading;
  }, [email, password, isLoading]);

  return (
    <main>
      {error ? <ErrorAlert>{error.message}</ErrorAlert> : null}
      {isLoading ? (
        <Spinner data-testid="spinner" animation="border" variant="primary" />
      ) : null}
      <Form onSubmit={handleSubmit}>
        <Form.Group className="mb-3" controlId="formBasicEmail">
          <Form.Label>Email address</Form.Label>
          <Form.Control
            name="email"
            value={email}
            onChange={handleChange}
            type="email"
            placeholder="Enter email"
          />
          <Form.Text className="text-muted">
            We'll never share your email with anyone else.
          </Form.Text>
        </Form.Group>

        <Form.Group className="mb-3" controlId="formBasicPassword">
          <Form.Label>Password</Form.Label>
          <Form.Control
            type="password"
            placeholder="Password"
            name="password"
            value={password}
            onChange={handleChange}
          />
        </Form.Group>
        <Form.Group className="mb-3" controlId="formBasicCheckbox">
          <Form.Check
            type="checkbox"
            label="Remember me!"
            name="remember"
            value="remember"
            checked={remember}
            onChange={handleChange}
          />
        </Form.Group>
        <Button variant="primary" type="submit" disabled={disabledButton}>
          Submit
        </Button>
      </Form>
    </main>
  );
}

export default LoginPage;
