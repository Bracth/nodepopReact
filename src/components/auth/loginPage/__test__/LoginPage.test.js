import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";
import LoginPage from "../LoginPage";
import { authLogin } from "../../../../store/actions";

jest.mock("../../../../store/actions");

describe("Login page", () => {
  test("snapshot", () => {
    const state = {
      ui: {
        error: null,
        isLoading: false,
      },
    };
    const store = {
      getState: () => state,
      dispatch: jest.fn(),
      subscribe: () => {},
    };
    const { container } = render(
      <Provider store={store}>
        <LoginPage />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
  test("should authLogin action", () => {
    authLogin.mockReturnValue("action");
    const email = "bryan@gmail.com";
    const password = "string";
    const state = {
      ui: {
        error: null,
        isLoading: false,
      },
    };

    const store = {
      getState: () => state,
      dispatch: jest.fn(),
      subscribe: () => {},
    };
    render(
      <Provider store={store}>
        <LoginPage />
      </Provider>
    );
    const emailInput = screen.getByLabelText(/Email/);
    const passwordInput = screen.getByLabelText(/Password/);
    const submitButton = screen.getByRole("button");
    console.log(submitButton);

    expect(submitButton).toBeDisabled();
    fireEvent.change(emailInput, { target: { value: email } });
    fireEvent.change(passwordInput, { target: { value: password } });
    expect(submitButton).not.toBeDisabled();

    fireEvent.click(submitButton);
    const credentials = authLogin.mock.calls[0][0];
    expect(credentials).toMatchObject({ email, password });
    expect(store.dispatch).toHaveBeenCalledWith("action");
  });

  test("should show an error with the message", () => {
    const error = { message: "Network error" };
    const state = {
      ui: {
        error,
        isLoading: false,
      },
    };
    const store = {
      getState: () => state,
      dispatch: jest.fn(),
      subscribe: () => {},
    };
    render(
      <Provider store={store}>
        <LoginPage />
      </Provider>
    );

    expect(screen.getByText(error.message)).toBeInTheDocument();
  });
  test("should show a spinner component when isLoading is true", () => {
    const state = {
      ui: {
        error: null,
        isLoading: true,
      },
    };

    const store = {
      getState: () => state,
      dispatch: jest.fn(),
      subscribe: () => {},
    };
    render(
      <Provider store={store}>
        <LoginPage />
      </Provider>
    );

    expect(screen.getByTestId("spinner")).toBeInTheDocument();
  });
});
