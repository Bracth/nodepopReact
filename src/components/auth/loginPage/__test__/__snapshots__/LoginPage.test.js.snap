// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page snapshot 1`] = `
<div>
  <main>
    <form
      class=""
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="formBasicEmail"
        >
          Email address
        </label>
        <input
          class="form-control"
          id="formBasicEmail"
          name="email"
          placeholder="Enter email"
          type="email"
          value=""
        />
        <small
          class="text-muted form-text"
        >
          We'll never share your email with anyone else.
        </small>
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="formBasicPassword"
        >
          Password
        </label>
        <input
          class="form-control"
          id="formBasicPassword"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      <div
        class="mb-3"
      >
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="formBasicCheckbox"
            name="remember"
            type="checkbox"
            value="remember"
          />
          <label
            class="form-check-label"
            for="formBasicCheckbox"
            title=""
          >
            Remember me!
          </label>
        </div>
      </div>
      <button
        class="btn btn-primary"
        disabled=""
        type="submit"
      >
        Submit
      </button>
    </form>
  </main>
</div>
`;
